plugin-server/src/worker/ingestion/properties-updater.ts

```diff
@@ -0,0 +1,122 @@
+import { Properties } from '@posthog/plugin-scaffold'
+import { DateTime } from 'luxon'
+import { QueryResult } from 'pg'
+
+import { Person, PersonPropertyUpdateOperation, TeamId } from '../../types'
+import { DB } from '../../utils/db/db'
+import { generateKafkaPersonUpdateMessage } from '../../utils/db/utils'
+
+export async function updatePersonProperties(
+    db: DB,
+    teamId: TeamId,
+    distinctId: string,
+    properties: Properties,
+    propertiesOnce: Properties,
+    timestamp: DateTime
+): Promise<void> {
+    if (Object.keys(properties).length === 0 && Object.keys(propertiesOnce).length === 0) {
+        return
+    }
+
+    let person: Person | undefined
+    await db.postgresTransaction(async (client) => {
+        person = await db.fetchPerson(teamId, distinctId, client, true)
+        if (!person) {
+            throw new Error(
+                `Could not find person with distinct id "${distinctId}" in team "${teamId}" to update props`
+            )
+        }
+
+        const shouldUpdate = calculateUpdatedProperties(person, properties, propertiesOnce, timestamp)
+        if (!shouldUpdate) {
+            return
+        }
+
+        const updateResult: QueryResult = await client.query(
+            `UPDATE posthog_person SET
+                properties = $1,
+                properties_last_updated_at = $2,
+                properties_last_operation = $3,
+                version = COALESCE(version, 0)::numeric + 1
+            WHERE id = $4
+            RETURNING version`,
+            [
+                JSON.stringify(person.properties),
+                JSON.stringify(person.properties_last_updated_at),
+                JSON.stringify(person.properties_last_operation || {}),
+                person.id,
+            ]
+        )
+        person.version = Number(updateResult.rows[0].version)
+    })
+
+    if (db.kafkaProducer && person) {
```

Comment:
Note: This line is pretty odd. We always update kafka even if person properties are not updated? Smells like a bug.

Reply:
That's just there to make the type checker happy we're always guaranteed to have a person at this point as if we didn't get a person we would have thrown at line 25. Open to any suggestions how to make this more easily readable/less confusing.

Reply:
right yes that's a bug (I thought line 32 handled it, but obviously it doesn't), will put up a PR